lib/src/cli.rs

```diff
@@ -609,16 +609,16 @@ async fn upgrade(opts: UpgradeOpts) -> Result<()> {
         }
     } else {
         let fetched = crate::deploy::pull(repo, imgref, None, opts.quiet).await?;
-        let staged_digest = staged_image.as_ref().map(|s| s.image_digest.as_str());
-        let fetched_digest = fetched.manifest_digest.as_str();
+        let staged_digest = staged_image.and_then(|s| s.digest().ok());
```

Comment:
I think we should either `?` or `expect("valid digest in status")` here but not `.ok()` which will silently discard invalid digests.

It really feels like it Shouldn't Happen and if it does we want a clear error.

Reply:
(There's a hugely complex subthread one could spawn from this around "how robust are we to corrupted on disk state and how hard is it to recover" but I think the most likely fix for that is basically ensuring that we can always pull a new image even if for some reason we can't read the current one)

Reply:
The rationale to `?` specifically is we already error out if e.g. the manifest JSON fails to parse because the disk state is corrupted, and "malformed digest" is basically that - the *writing* code shouldn't have let it happen, and we're treating "on disk state is malformed" as an error (but not a crash/panic).

Reply:
:+1: I went the `expect` route since we want to be explicit that errors there are... unexpected :sunglasses: